### PR TITLE
Added Documentation for pyomo examples (batch_processing.py)

### PR DESCRIPTION
**Pull request by AlbertLee125**
I added the documentation for the batch_processing.py.

### CODE REVIEW

**Comment by AlbertLee125** [comment]
From the TODO list, I can assume the code is formulated based on the existing GAMS studio code. I will remove the TODO and then rewrite the documentation.

File: gdplib/pyomo_examples/batch_processing.py [hunk 1]
```diff
@@ -23,7 +32,7 @@ def build_model():
     model.BigM[None] = 1000
 
 
-    ## Constants from GAMS
+    # Constants from GAMS
     StorageTankSizeFactor = 2*5 # btw, I know 2*5 is 10... I don't know why it's written this way in GAMS?
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py]: Myabe you can remove this comment and just write 10

File: gdplib/pyomo_examples/batch_processing.py [hunk 2]
```diff
@@ -36,10 +45,7 @@ def build_model():
     # TODO: YOU ARE HERE. YOU HAVEN'T ACTUALLY MADE THESE THE BOUNDS YET, NOR HAVE YOU FIGURED OUT WHOSE
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:68]: Is this TODO still required?

File: gdplib/pyomo_examples/batch_processing.py [hunk 3]
```diff
@@ -52,12 +58,9 @@ def filter_out_last(model, j):
 
 
     # TODO: these aren't in the formulation??
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:100]: Can you review this?

**Review by bernalde — CHANGES_REQUESTED** [review]

**Comment by bernalde** [comment]
Do it

File: gdplib/pyomo_examples/batch_processing.py (hunk 1 above)
Comment (bernalde) [reply gdplib/pyomo_examples/batch_processing.py]: You resolved the comments without addressing it. Please work on it and commit it before resolving the comments

File: gdplib/pyomo_examples/batch_processing.py (hunk 2 above)
Comment (bernalde) [reply gdplib/pyomo_examples/batch_processing.py:68]: Same as above

File: gdplib/pyomo_examples/batch_processing.py (hunk 3 above)
Comment (bernalde) [reply gdplib/pyomo_examples/batch_processing.py:100]: Same as above

File: gdplib/pyomo_examples/batch_processing.py [hunk 4]
```diff
@@ -138,21 +139,15 @@ def get_storageTankSize_bounds(model, j):
...
             model.Alpha2 * sum(exp(model.Beta2 * model.storageTankSize_log[j]) for j in model.STAGESExceptLast)
     model.min_cost = Objective(rule=get_cost_rule)
 
-
-    ##############
     # Constraints
-    ##############
-
     def processing_capacity_rule(model, j, i):
         return model.volume_log[j] >= log(model.ProductSizeFactor[i, j]) + model.batchSize_log[i, j] - \
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:308]: Function documentation missing

File: gdplib/pyomo_examples/batch_processing.py [hunk 5]
```diff
@@ -169,9 +164,7 @@ def finish_in_time_rule(model):
     model.finish_in_time = Constraint(rule=finish_in_time_rule)
 
 
-    ###############
     # Disjunctions
-    ###############
 
     def storage_tank_selection_disjunct_rule(disjunct, selectStorageTank, j):
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:357]: Function documentation missing

**Review by bernalde — CHANGES_REQUESTED** [review]
Please address the comments before resolving the conversations

File: gdplib/pyomo_examples/batch_processing.py [hunk 6]
```diff
@@ -24,7 +24,13 @@
 
 
 def build_model():
+    """_summary_
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py]: Please do not commit this kind of changes. Commits should have more meaningful states

**Review by bernalde — CHANGES_REQUESTED** [review]
Do not do commits with intermediate states unless you are working on "work in progress" PRs [WIP]

File: gdplib/pyomo_examples/batch_processing.py (hunk 2 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:68]: I think I will just leave it as it was.

File: gdplib/pyomo_examples/batch_processing.py (hunk 6 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py]: I added the documentation for each function definition.

File: gdplib/pyomo_examples/batch_processing.py (hunk 5 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:357]: Added the Documentation of the functions.

File: gdplib/pyomo_examples/batch_processing.py (hunk 4 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:308]: Added Function documentation.

File: gdplib/pyomo_examples/batch_processing.py [hunk 7]
```diff
@@ -4,27 +4,49 @@
 from pyomo.environ import *
 from pyomo.gdp import *
 
-'''Problem from http://www.minlp.org/library/problem/index.php?i=172&lib=GDP
-We are minimizing the cost of a design of a plant with parallel processing units and storage tanks
-in between. We decide the number and volume of units, and the volume and location of the storage
-tanks. The problem is convexified and has a nonlinear objective and global constraints
+# Problem from http://www.minlp.org/library/problem/index.php?i=172&lib=GDP
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:17]: The other codes had a header from Pyomo. Please include that here

File: gdplib/pyomo_examples/batch_processing.py [hunk 8]
```diff
@@ -4,27 +4,49 @@
...
+
+    Returns
+    -------
+    Pyomo.ConcreteModel
+        An instance of the Pyomo ConcreteModel class representing the batch processing optimization model, 
+        ready to be solved with an appropriate solver.
+
+    References
+    ----------
+    Ravemark, E. Optimization models for design and operation of chemical batch processes. Ph.D. Thesis, ETH Zurich, 1995.
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py]: DOI missing

File: gdplib/pyomo_examples/batch_processing.py [hunk 9]
```diff
@@ -4,27 +4,49 @@
...
 
     # TODO: it looks like they set a bigM for each j. Which I need to look up how to do...
     model.BigM = Suffix(direction=Suffix.LOCAL)
     model.BigM[None] = 1000
 
 
-    ## Constants from GAMS
-    StorageTankSizeFactor = 2*5 # btw, I know 2*5 is 10... I don't know why it's written this way in GAMS?
+    # Constants from GAMS
+    StorageTankSizeFactor = 10
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:59]: Do you have any details on the units or a more detailed description of these?

File: gdplib/pyomo_examples/batch_processing.py [hunk 10]
```diff
@@ -79,22 +112,36 @@ def filter_out_last(model, j):
...
         return log(model.PARALLELUNITS.ord(k))
 
     model.LogCoeffs = Param(model.PARALLELUNITS, initialize=get_log_coeffs)
 
     # bounds
-    model.volumeLB = Param(model.STAGES, default=VolumeLB)
-    model.volumeUB = Param(model.STAGES, default=VolumeUB)
+    model.volumeLB = Param(model.STAGES, default=VolumeLB, doc='Lower Bound of Volume of the Units')
+    model.volumeUB = Param(model.STAGES, default=VolumeUB, doc='Upper Bound of Volume of the Units')
     model.storageTankSizeLB = Param(model.STAGES, default=StorageTankSizeLB)
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py]: Missing docs

File: gdplib/pyomo_examples/batch_processing.py [hunk 11]
```diff
@@ -119,17 +166,80 @@ def get_log_coeffs(model, k):
...
+        j : int
+            Index for the processing stages in the plant. Stages are ordered and include various processing steps.
+
+        Returns
+        -------
+        tuple
+            A tuple containing the lower and upper bounds for the volume of processing units at stage j..
+        """
         return (model.volumeLB[j], model.volumeUB[j])
     model.volume_log = Var(model.STAGES, bounds=get_volume_bounds)
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py]: Missing docs

File: gdplib/pyomo_examples/batch_processing.py [hunk 12]
```diff
@@ -119,17 +166,80 @@ def get_log_coeffs(model, k):
...
+        j : int
+            Index for the processing stages in the plant. Stages are ordered and include various processing steps.
+
+        Returns
+        -------
+        tuple
+            A tuple containing the minimum and maximum bounds for the logarithmic number of units in phase at stage j, ensuring model constraints are met.
+        """
         return (0, model.unitsInPhaseUB[j])
     model.unitsInPhase_log = Var(model.STAGES, bounds=get_unitsInPhase_bounds)
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py]: Missing docs

File: gdplib/pyomo_examples/batch_processing.py [hunk 13]
```diff
@@ -202,27 +459,111 @@ def no_batch_rule(disjunct, i):
...
+
+        Returns
+        -------
+        list
+            A list of disjuncts representing the choices for including or not including a storage tank between stages j and j+1.
+        """
         return [model.storage_tank_selection_disjunct[selectTank, j] for selectTank in [0,1]]
     model.select_storage_tanks = Disjunction(model.STAGESExceptLast, rule=select_storage_tanks_rule)
 
     # though this is a disjunction in the GAMs model, it is more efficiently formulated this way:
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:490]: What does this comment refer to?

File: gdplib/pyomo_examples/batch_processing.py [hunk 14]
```diff
@@ -202,27 +459,111 @@ def no_batch_rule(disjunct, i):
...
+
+        Notes
+        -----
+        These are not directly related to disjunctions but more to the logical modeling of unit operations.
+        """
         return model.unitsInPhase_log[j] == sum(model.LogCoeffs[k] * model.inPhase[j,k] \
                                                 for k in model.PARALLELUNITS)
     model.units_in_phase = Constraint(model.STAGES, rule=units_in_phase_rule)
 
     # and since I didn't do the disjunction as a disjunction, we need the XORs:
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py]: This comment is no longer relevant

**Review by bernalde — CHANGES_REQUESTED** [review]
This file still needs work. @tristantc please aid Albert in it. You might need to go back to the GAMS implementation described here https://www.minlplib.org/batchs101006m.html and found here http://egon.cheme.cmu.edu/ibm/page.htm and here https://minlp.org/library/problem/index.php?i=172&lib=GDP

File: gdplib/pyomo_examples/batch_processing.py (hunk 3 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:100]: The one written in the TODO list. 
https://www.minlp.org/library/problem/index.php?i=172&lib=GDP I checked the Nomenclature of the Design of the Multi-product Batch Plant. There are no sets for the storage tanks.

File: gdplib/pyomo_examples/batch_processing.py (hunk 14 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py]: I will remove the comments.

File: gdplib/pyomo_examples/batch_processing.py (hunk 7 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:17]: Added the header.

File: gdplib/pyomo_examples/batch_processing.py (hunk 9 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:59]: Unfortunately, I do not know the units. https://www.minlp.org/problems/mod/229/GDP_model.pdf According to this website. They only show the nomenclature not the detail unit.

Comment (bernalde) [reply gdplib/pyomo_examples/batch_processing.py:59]: Not there, but go deeper. Check Ravemach thesis

Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:59]: https://www.research-collection.ethz.ch/handle/20.500.11850/142491
I cannot get the access for the Full text.

Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:59]: Here is the DOI of his. https://doi.org/10.3929/ethz-a-001591449
I can only get the access of the Abstract. Is there any way to get approval for full text?

File: gdplib/pyomo_examples/batch_processing.py [hunk 15]
```diff
@@ -242,11 +252,11 @@ def get_storageTankSize_bounds(model, j):
         """
         return (model.storageTankSizeLB[j], model.storageTankSizeUB[j])
     # TODO: these bounds make it infeasible...
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:254]: Does this mean this function is not used? If not, please add more context to the comment

File: gdplib/pyomo_examples/batch_processing.py (hunk 13 above)
Comment (bernalde) [reply gdplib/pyomo_examples/batch_processing.py:490]: @AlbertLee125 @tristantc @ZedongPeng

**Review by bernalde — CHANGES_REQUESTED** [review]
Some comments are still open.
We should also start adding the description of the files in the README in a next PR

File: gdplib/pyomo_examples/batch_processing.py (hunk 8 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py]: Added the DOI

File: gdplib/pyomo_examples/batch_processing.py (hunk 15 above)
Comment (AlbertLee125) [reply gdplib/pyomo_examples/batch_processing.py:254]: I think what the TODO: was talking about setting the bound, or hard coding the boundary can make the boundary infeasible. Shall I erase it?

File: gdplib/pyomo_examples/batch_processing.py [hunk 16]
```diff
@@ -108,14 +108,14 @@ def filter_out_last(model, j):
...
+    model.ProcessingTime = Param(model.PRODUCTS, model.STAGES, doc='Processing Time')
 
     # These are hard-coded in the GAMS file, hence the defaults
-    model.StorageTankSizeFactor = Param(model.STAGES, default=StorageTankSizeFactor)
+    model.StorageTankSizeFactor = Param(model.STAGES, default=StorageTankSizeFactor, doc='Storage Tank Size Factor')
     model.StorageTankSizeFactorByProd = Param(model.PRODUCTS, model.STAGES,
-                                              default=StorageTankSizeFactorByProd)
+                                              default=StorageTankSizeFactorByProd, doc='Storage Tank Size Factor by Product')
 
     # TODO: bonmin wasn't happy and I think it might have something to do with this?
```
Comment (bernalde) [inline gdplib/pyomo_examples/batch_processing.py:120]: This is an interesting comment. I will not keep this PR open but we should take this into account.

**Review by bernalde — APPROVED** [review]

**Pull request merged**